### DISCUSSION TITLE
[HttpClient] How to modify the request and add a signature header just before sending it?

### DISCUSSION THREAD

**Discussion by 1ed** [symfony/symfony#52608] [Q&A]
Hi, I need to add some signature to the request headers before I send the request. The signature calculated based on the request method, URL, headers and the body. Is there a way to decorate the Symfony client or the response and access these to calculate and add the signature to the request **just before sending** it? I mean I can access the options ant others, but the the client does a lot of stuff before actually sending the request, like json encoding, adding headers etc. With a PSR client this can be done easily, but there is a way to do it with the others implementing `Symfony\Contracts\HttpClient\HttpClientInterface`? Thanks!

**Comment by nicolas-grekas** [comment]
You gave the answer in the question : you should write a decorator. You can leverage `DecoratorTrait` to make it a bit easier.

**Reply by 1ed** [reply]
Thank you for your answer, Nicolas!

Yes, but how can I access the request right before the decorated client sends it? Here are some code:

```php
class SigningHttpClient implements HttpClientInterface
{
    use DecoratorTrait;

    // ...

    public function request(string $method, string $url, array $options = []): ResponseInterface
    {
        // I can get these from options, but the decorated client does a lot of logic
        // based on the options and add new headers, creates the body from json, ...
        $headers = $options['headers'] ?? [];
        $body = $options['body'] ?? '';

        // this service needs all the headers and the final body of the request
        $signatures = $this->signer->sign($options['extra']['secret_key'], $method, $url, $headers, $body);
        unset($options['extra']['secret_key']);
        $options['headers'] = [...$options['headers'], ...$signatures];

        return $this->client->request($method, $url, $options);
    }
}

```
Maybe I could use `prepareRequest` and other methods from `HttpClientTrait`, but AFAIK the different client implementations have logic that can modify the request before they actually sending it and in that case the signature will be wrong. Can I somehow get the final headers and body just before the decorated client sends it, without copying a client implementation, or it is not possible? I was wondering that decoration is not a good pattern for this use case and maybe I should use the response and `AsyncResponse` somehow, but that didn't lead anywhere either.

**Reply by nicolas-grekas** [reply]
You need to ensure that your decorator is registered as the last one - just around the real client. Whatever the way to plug into that stack (middleware, listeners, decorators), the priority will always be something you'll need to configure.
Then, yes, the real client can also add some headers. But it won't add them if you already defined them before. If you need to sign only a subset of the headers, then that's going to be easier.
Note that this requirement for signature is strange. HTTPS is a way better way to add security at this layer...
